Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,320 @@
+      · generalize_proofs; simp [*]
+    · simp only [category_comp_f, category_comp_φ, Category.assoc, limit.lift_π,
+        PullbackCone.mk_pt, PullbackCone.mk_π_app] at h₂₂ ⊢
+      convert h₂₂ i using 2
+      refine eq_of_heq ((eqToHom_comp_heq_iff _ _ _).2 ((heq_comp_eqToHom_iff _ _ _).2 ?_))
+      congr 1
+      · rw [hf]
+      · rw [hf]
+      · rw [hf]
+      · exact (comp_eqToHom_heq_iff _ _ _).2 ((heq_comp_eqToHom_iff _ _ _).2 (by rw [hf]))
+      · rw [hf]
+      · generalize_proofs; simp [*]
+
+
+section HasCoproducts
+
+variable [HasCoproducts.{w} A]
+
+variable (C) in
+@[simps] noncomputable def eval : (C ⥤ A) ⥤ (FormalCoproduct.{w} C ⥤ A) where
+  obj F :=
+    { obj X := ∐ fun (i : X.I) ↦ F.obj (X.obj i)
+      map {X Y} f := Sigma.desc fun i ↦ F.map (f.φ i) ≫ Sigma.ι (F.obj ∘ Y.obj) (f.f i)
+      map_comp _ _ := Sigma.hom_ext _ _ (fun _ ↦ by simp [Sigma.ι_desc]) }
+  map α := { app f := Sigma.map fun i ↦ α.app (f.obj i) }
+
+noncomputable def evalOf : eval C A ⋙ (whiskeringLeft _ _ A).obj (of C) ≅ Functor.id (C ⥤ A) :=
+  NatIso.ofComponents fun F ↦ NatIso.ofComponents
+    (fun x ↦ ⟨Sigma.desc fun _ ↦ 𝟙 _, Sigma.ι (fun _ ↦ F.obj x) PUnit.unit, by aesop, by simp⟩)
+    (fun f ↦ Sigma.hom_ext _ _ (by simp [Sigma.ι_desc]))
+
+variable {A} in
+@[simps!] noncomputable
+def evalCoproductIso (F : C ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    ((eval C A).obj F).obj (coproduct J f).cocone.pt ≅ ∐ fun j ↦ ((eval C A).obj F).obj (f j) :=
+  (sigmaSigmaIso (fun j ↦ (f j).I) (fun j x ↦ F.obj ((f j).obj x))).symm
+
+lemma comp_evalCoproductIso (F : C ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C)
+    (j : J) (x : (f j).I) :
+    ((eval C A).obj F).map ((ofHomEquiv ((f j).obj x) _).2 ⟨Sigma.mk j x, 𝟙 _⟩) ≫
+        (evalCoproductIso F f).hom =
+      Sigma.desc fun _ ↦ Sigma.ι (fun i ↦ F.obj ((f j).obj i)) x ≫
+        Sigma.ι (fun b ↦ ∐ fun i ↦ F.obj ((f b).obj i)) j :=
+  Sigma.hom_ext _ _ fun _ ↦ by simp [ofHomEquiv, Sigma.ι_desc]
+
+lemma sigmaComparison_eq (F : C ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    sigmaComparison ((eval C A).obj F) f =
+      (evalCoproductIso F f).inv ≫ ((eval C A).obj F).map (coproductIsoCoproduct f).inv := by
+  refine Sigma.hom_ext _ _ fun j ↦ Sigma.hom_ext _ _ fun x ↦ ?_
+  rw [ι_comp_sigmaComparison, ← Functor.mapIso_inv, ← Category.assoc, ← Category.assoc,
+    Iso.eq_comp_inv, evalCoproductIso, Iso.symm_inv, sigmaSigmaIso_hom, Category.assoc,
+    Category.assoc, Sigma.ι_desc, Sigma.ι_desc, Functor.mapIso_hom, ← Functor.map_comp,
+    coproductIsoCoproduct_hom, colimit.ι_desc]
+  simp [coproduct]
+  rfl
+
+variable {A} in
+theorem preservesCoproductEval (F : C ⥤ A) {J : Type w} (f : J → FormalCoproduct.{w} C) :
+    PreservesColimit (Discrete.functor f) ((eval.{w} C A).obj F) := by
+  refine CategoryTheory.Limits.PreservesCoproduct.of_iso_comparison _ _ (i := ?_)
+  rw [sigmaComparison_eq, ← Functor.mapIso_inv, ← Iso.trans_inv, ← Iso.symm_hom]
+  infer_instance
```

Comment:
Not fully tested, but perhaps this can be simplified in the following way: 
if you define 
```
def isColimitEvalMapCocone (𝒜 : Type w) (f : 𝒜 → FormalCoproduct.{w} C) (F : C ⥤ A) : 
    IsColimit (((eval.{w} C A).obj F).mapCocone (coproductCocone.{w} (C := C) 𝒜 f)) := 
  sorry
```
directly, which should hopefully be doable given that this mapCone is very close to the coproduct cocone in `A` for the right set of objects. Then you get basically all the isos here as part of the "unicity of colimits", and you also can directly show that the functor preserves coproducts more easily, since this is defined as some `Nonempty (((eval.{w} C A).obj F).mapCocone _)`.